File: flat2tree.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class flat2tree {
    public static void main( String[] args) {
        List<TreeNode> flat = new ArrayList<TreeNode>() {
            {
                add(new TreeNode(1, null, "data"));
                add(new TreeNode(2, null, "data"));
                add(new TreeNode(3, 1, "data"));
                add(new TreeNode(4, 1, "data"));
                add(new TreeNode(5, 3, "data"));
                add(new TreeNode(6, 2, "data"));
            }
        };

        List<TreeNode> tree = parse(flat);
    }

    public static List<TreeNode> parse(List<TreeNode> flat) {
        return parse(flat, null, new ArrayList<TreeNode>());
    }

    public static List<TreeNode> parse(List<TreeNode> flat, TreeNode parent , List<TreeNode> tree  ){
        Integer parentid = parent != null ? parent.id : null;
        List<TreeNode> children = flat.stream().filter(child -> child.parentid == parentid).collect(Collectors.toList());
        if (children.size() != 0) {
            if (parentid != null) {
                parent.children = children;
            } else {
               tree.addAll(children);
            }
            children.forEach(child -> parse(flat,child,tree));
        }
        return tree;
    }

}

class TreeNode {

    public List<TreeNode> children;

    public Integer id;
    public Integer parentid;
    public String data;

    public TreeNode(Integer id, Integer parentid, String data) {
        this.id = id;
        this.parentid = parentid;
        this.data = data;

    }
}

File: flat2tree.js

"use strict"

class Node {
    constructor(id, parentid, data) {
        this.id = id
        this.parentid = parentid
        // this.parent = undefined
        this.children = []
        this.data = data
    }
    setParent(parent) {
        this.parent = parent
    }

    isLeaf() {
        return !!this.children.length
    }

}



class Tree {

    constructor() {
        this.data = []
        this.simple = []
        this.flat = []
    }
    getPath(node) {
        let path = []
        while (node != null) {
            path.push(node.id)
            node = this.flat.find(x => x.id == node.parentid)
        }
        return path
    }
    traverse(operation, node) {
        if (!node) {
            this.data.forEach(n => this.traverse(operation, n))
        } else if (node.children.length) {
            node.children.forEach(n => this.traverse(operation, n))
            operation(node)
        }
        else {
            operation(node)
        }
    }

    getNode(id) {
        return this.flat.find(x => x.id == id)
    }
    removeNode(node) {
        var index = this.flat.indexOf(x => x[this.config.idProperty] == node[this.config.idProperty])
        if (index != -1) {
            this.flat.splice(index, 1);
        }
    }

    toJSON() {
        return JSON.stringify(this.data, null, 4)
    }

    parse(array, { dataProperty = 'data', idProperty = 'id', parentidProperty = 'parentid' } = {}) {
        this.config = { dataProperty, idProperty, parentidProperty }
        this.data = []
        this.flat = array.map(x => new Node(x[idProperty], x[parentidProperty], x[dataProperty]))
        this.parseRecursive(null)
    }
    parseRecursive(parent) {
        let parentid = parent ? parent[this.config.idProperty] : null
        var children = this.flat.filter(child => child[this.config.parentidProperty] == parentid);

        if (children.length) {
            if (parentid) {
                // children.forEach(child => child.setParent(parent))
                parent.children = children;
            } else {
                this.data = children
            }
            children.forEach(child => this.parseRecursive(child))
        }
        return this;
    }
}

module.exports = { Tree,Node }